### PR TITLE
fJVT scale factors support

### PR DESCRIPTION
**Pull request by ntadej**
This PR fixes fJVT when using systematics and adds fJVT scale factors.

It might be good idea for the future to move fJVT to `JetCalibrator` to avoid the need of deep copying.

### CODE REVIEW

File: Root/JetSelector.cxx [hunk 1]
```diff
@@ -236,13 +236,47 @@ EL::StatusCode JetSelector :: initialize ()
...
 
   //init fJVT
-  ANA_MSG_DEBUG("Trying to initialize fJVT tool");
-  //ANA_CHECK( ASG_MAKE_ANA_TOOL(m_fJVT_tool_handle, JetForwardJvtTool));
-  ANA_CHECK(m_fJVT_tool_handle.retrieve());
-  ANA_MSG_DEBUG("Successfully initialized fJVT tool");
+  if (m_dofJVT) {
+    ANA_MSG_DEBUG("Trying to initialize fJVT tool");
+    //ANA_CHECK( ASG_MAKE_ANA_TOOL(m_fJVT_tool_handle, JetForwardJvtTool));
+    ANA_CHECK(m_fJVT_tool_handle.retrieve());
```
Comment (kratsg) [inline Root/JetSelector.cxx]: missing a setToolName call here (we need it to make sure the tool is effective private)

File: Root/JetSelector.cxx [hunk 2]
```diff
@@ -236,13 +236,47 @@ EL::StatusCode JetSelector :: initialize ()
...
+    ANA_MSG_DEBUG("Successfully initialized fJVT tool");
+
+    // initialize the CP::JetJvtEfficiency Tool for fJVT
+    setToolName(m_fJVT_eff_tool_handle, "fJVT_eff_tool");
```
Comment (kratsg) [inline Root/JetSelector.cxx]: Don't need to set the tool name here (no second argument) unless there's more than one tool in this algorithm for it, or if you need to share it with another algorithm

File: Root/JetSelector.cxx [hunk 3]
```diff
@@ -236,13 +236,47 @@ EL::StatusCode JetSelector :: initialize ()
...
+        break;
+      }
+      ANA_MSG_INFO("\t " << syst_it.name());
+    }
+  }
 
-  // initialize the CP::JetJvtEfficiency Tool
-  setToolName(m_JVT_tool_handle);
+  // initialize the CP::JetJvtEfficiency Tool for JVT
+  setToolName(m_JVT_tool_handle, "JVT_eff_tool");
```
Comment (kratsg) [inline Root/JetSelector.cxx]: don't need to set name unless you need to share with another algorithm.

File: Root/JetSelector.cxx [hunk 4]
```diff
@@ -614,6 +649,109 @@ bool JetSelector :: executeSelection ( const xAOD::JetContainer* inJets,
...
+              }
+            }
+          }
+          //
+          // Add it to decoration vector
+          //
+          sfVecfJVT( *jet ).push_back( fjvtSF );
+
+          ANA_MSG_DEBUG( "===>>>");
+          ANA_MSG_DEBUG( " ");
```
Comment (kratsg) [inline Root/JetSelector.cxx]: no need for empty line

File: Root/JetSelector.cxx [hunk 5]
```diff
@@ -614,6 +649,109 @@ bool JetSelector :: executeSelection ( const xAOD::JetContainer* inJets,
...
+          ANA_MSG_DEBUG( "Jet " << idx << ", pt = " << jet->pt()*1e-3 << " GeV, |eta| = " << std::fabs(jet->eta()) );
+          ANA_MSG_DEBUG( " ");
```
Comment (kratsg) [inline Root/JetSelector.cxx]: empty line

File: Root/JetSelector.cxx [hunk 6]
```diff
@@ -614,6 +649,109 @@ bool JetSelector :: executeSelection ( const xAOD::JetContainer* inJets,
...
+          ANA_MSG_DEBUG( "fJVT SF decoration: " << m_outputSystNamesfJVT );
+          ANA_MSG_DEBUG( " ");
```
Comment (kratsg) [inline Root/JetSelector.cxx]: empty line

File: Root/JetSelector.cxx [hunk 7]
```diff
@@ -614,6 +649,109 @@ bool JetSelector :: executeSelection ( const xAOD::JetContainer* inJets,
...
+          ANA_MSG_DEBUG( "Systematic: " << syst_it.name() );
+          ANA_MSG_DEBUG( " ");
```
Comment (kratsg) [inline Root/JetSelector.cxx]: empty line

File: xAODAnaHelpers/JetSelector.h [hunk 8]
```diff
@@ -165,6 +165,38 @@ class JetSelector : public xAH::Algorithm
...
+
+        The corresponding data/MC SF will be saved as a ``std::vector<float>`` decoration (for MC only), for nominal WP and the available systematics.
+
+        ======== ============== =============
+        Value    HS Efficiency  PU Fake Rate
+        ======== ============== =============
+        "Medium"  87.1-97.0%     53.4-60.9%
+        "Tight"   79.9-95.6%     45.4-50.3%
+        ======== ============== =============
+
```
Comment (kratsg) [inline xAODAnaHelpers/JetSelector.h]: can you add a link to where the above numbers are from?

File: xAODAnaHelpers/JetSelector.h [hunk 9]
```diff
@@ -165,6 +165,38 @@ class JetSelector : public xAH::Algorithm
...
+    @endrst
+  */
+  std::string m_WorkingPointfJVT = "Medium";
+
+  /**
+     @brief Configuration containting fJVT scale factors.
+
+     @rst
+     The configuration file with the scale factors calculated by the ``CP::IJetJvtEfficiency``.
```
Comment (kratsg) [inline xAODAnaHelpers/JetSelector.h]: the text inside `@rst/..../@endrst` needs to be indented at least once.

File: Root/JetSelector.cxx (hunk 2 above)
Comment (ntadej) [inline Root/JetSelector.cxx]: There are two instances of tools, one for JVT and one for fJVT. If I do not set the names, the same one is used both times (the one which is initialised first).

Comment (kratsg) [inline Root/JetSelector.cxx]: Good point. Wasn't obvious at first but now I see it.

**Comment by ntadej** [comment]
I've also noticed that we need passed info for data so I added it. I'm open for suggestions where to put accessors in JetContainer now.

**Comment by kratsg** [comment]
> I've also noticed that we need passed info for data so I added it. I'm open for suggestions where to put accessors in JetContainer now.

This seems fine for right now. We might want to provide a uniform area for all accessors in our code, but I suspect this is a long term goal / cleanup effort.

**Comment by ntadej** [comment]
I addressed all of your comments. I can rebase if needed.

The only thing that is still on my mind is to check if deep copies are really necessary for fJVT, or should I open a new MR for it?

**Comment by ntadej** [comment]
I updated branches to be of type `std::vector<int>` not `std::vector<char>` + rebased.

Also it seems that fJVT in current place needs deep copies so I will try to fix that in a separate PR.

**Comment by ntadej** [comment]
Hi @kratsg, is there anything else that needs to be done from my side for this to be merged?

**Comment by kratsg** [comment]
> Hi @kratsg, is there anything else that needs to be done from my side for this to be merged?

Nope, but thanks for poking me. I've merged it in :)

**Comment by kratsg** [comment]
I also approved the changes (but forgot to mark that on the review, sorry).

**Pull request merged**